prober/derp.go

```diff
@@ -295,6 +319,192 @@ func (d *derpProber) probeBandwidth(from, to string, size int64) ProbeClass {
+				prometheus.MustNewConstHistogram(prometheus.NewDesc("derp_qd_probe_delays_seconds", "Distribution of queuing delays", nil, l), qdh.count, qdh.sum, qdh.buckets),
+			}
+			qdh.mx.Unlock()
+			return result
+		},
+	}
+}
+
+// derpProbeQueuingDelay continuously sends data between two local DERP clients
+// connected to two DERP servers in order to measure queuing delays.
+func derpProbeQueuingDelay(ctx context.Context, dm *tailcfg.DERPMap, from, to *tailcfg.DERPNode, packetsDropped *expvar.Float, qdh *queuingDelayHistogram) (err error) {
+	// This probe uses clients with isProber=false to avoid spamming the derper logs with every packet
+	// sent by the bandwidth probe.
+	fromc, err := newConn(ctx, dm, from, false)
+	if err != nil {
+		return err
+	}
+	defer fromc.Close()
+	toc, err := newConn(ctx, dm, to, false)
+	if err != nil {
+		return err
+	}
+	defer toc.Close()
+
+	// Wait a bit for from's node to hear about to existing on the
+	// other node in the region, in the case where the two nodes
+	// are different.
+	if from.Name != to.Name {
+		time.Sleep(100 * time.Millisecond) // pretty arbitrary
+	}
+
+	if err := runDerpProbeQueuingDelayContinously(ctx, from, to, fromc, toc, packetsDropped, qdh); err != nil {
+		// Record pubkeys on failed probes to aid investigation.
+		return fmt.Errorf("%s -> %s: %w",
+			fromc.SelfPublicKey().ShortString(),
+			toc.SelfPublicKey().ShortString(), err)
+	}
+	return nil
+}
+
+func runDerpProbeQueuingDelayContinously(ctx context.Context, from, to *tailcfg.DERPNode, fromc, toc *derphttp.Client, packetsDropped *expvar.Float, qdh *queuingDelayHistogram) error {
+	// Circular buffer to hold packet send times. It is sized to hold timings
+	// for up to 5 seconds when sending packets at a rate of 10 per second.
+	// It assumes that packets may be dropped, but that they will generally
+	// arrive in order. Packets arriving out of order will result in older
+	// packets being ignored, effectively overcounting the number of dropped
+	// packets.
+	sentTimes := circularqueue.NewFIFO(50, func(t time.Time) {
+		// If a sent time is evicted, that means we'll never record a timing
+		// for this packet, so we considered it dropped.
+		packetsDropped.Add(1)
+	})
+
+	// Send the packets.
+	sendErrC := make(chan error, 1)
+	pkt := make([]byte, 260) // the same size as a CallMeMaybe packet observed on a Tailscale client.
+	crand.Read(pkt)
+
+	go func() {
+		t := time.NewTicker(time.Second / 10) // 10 packets per second
+		defer t.Stop()
+
+		seq := 0
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case <-t.C:
+				sentTimes.Push(time.Now())
+				binary.BigEndian.PutUint64(pkt, uint64(seq))
+				seq++
+				if seq == math.MaxInt {
+					// We're extremely unlikely to hit this, but if we do, we should stop.
+					sendErrC <- errors.New("Packet sequence number exhausted")
```

Comment:
Missing a return here

Reply:
I decided to move this into the `FIFO` queue type as a panic, since it isn't able to handle this case either. This way, if someone else uses that type, they won't get bitten in the unlikely event of the index wrapping around.